src/runtime_src/core/pcie/linux/shim.cpp

```diff
@@ -131,28 +132,220 @@ namespace xocl {
+                    uint64_t priv_data)
+    {
+        iov[0].iov_base = header;
+        iov[0].iov_len = sizeof(struct xocl_qdma_req_header);
+        iov[1].iov_base = (void *)buf_va;
+        iov[1].iov_len = buf_len;
+
+        if (cb) {
+            memset(cb, 0, sizeof(struct iocb));
+            cb->aio_fildes = (int)qhndl;
+            cb->aio_lio_opcode = h2c ? IOCB_CMD_PWRITEV : IOCB_CMD_PREADV;
+            cb->aio_buf = (uint64_t)iov;
+            cb->aio_offset = 0;
+            cb->aio_nbytes = 2;
+            cb->aio_data = priv_data;
+        }
+    }
+
+    /* submit all of the queued i/o */
+    int queue_flush_aio_request(void)
+    {
+        reqLock.lock();
+
+        /* if there is submission error, wait till all requests are drained */
+        if (cbErrCnt) {
+            reqLock.unlock();
+            return 0;
+        }
+
+        do {
+            /* submit all queued requests */
+            unsigned int cb_max = bufCnt;
+            reqLock.unlock();
+
+            unsigned int cb_cnt = 0;
+            struct iocb **cbpp = (struct iocb **)malloc(cb_max * sizeof(struct iocb *));
+
+            reqLock.lock();
+            auto end = reqList.end();
+            for (auto it = reqList.begin(); it != end && cb_cnt < cb_max;
+                ++it, cb_cnt++) {
+                auto qio = (*it);
+
+                qio.header.flags = qio.flags;
+                prepare_io(&qio.cb, qio.iov, &qio.header, qio.buf_va, qio.len,
+				qio.priv_data);
+                *(cbpp+cb_cnt) = &qio.cb;
+            }
+
+            int submitted = io_submit(qAioCtx, cb_cnt, cbpp);
+            free(cbpp);
+            if (submitted < 0) {
+                if (submitted != -EAGAIN) {
+		    /* something went wrong, flush all of the queued request
+                     * with this error */
+                    cbErrCnt = reqList.size();
+                    cbErrCode= submitted;
+		}
+                reqLock.unlock();
+                return submitted;
+
+            } else if (submitted > 0) {
+                release_request(submitted);
+                cbSubmitCnt += submitted;
+            }
+        } while(reqList.size());
+        reqLock.unlock();
+
+        return 0;
+    }
+
+    /* worker thread for aio batching */
+    void queue_aio_worker(void)
+    {
+        byteCnt = bufCnt = 0;
+        cbSubmitCnt = cbPollCnt = cbErrCnt = 0;
+
+
+        std::mutex mtx;
+        std::unique_lock<std::mutex> lck(mtx);
+        do {
+	    /* set a time out to prevent holding the i/o forever */
+            if (!timer_ms)
+		cv.wait_for(lck, std::chrono::milliseconds(1000));
+            else
+		cv.wait_for(lck, std::chrono::milliseconds(timer_ms));
+
+            reqLock.lock();
+            if (reqList.empty()) {
+                reqLock.unlock();
+                continue;
+            }
+            reqLock.unlock();
+
+            if (qExit) break;
+
+            queue_flush_aio_request();
+          
+        } while(!qExit);
```

Comment:
why this test?  there is already a break on qExit, so I infer !qExit must always be true if tested.